app/code/Magento/CatalogUrlRewrite/Model/Products/AdaptUrlRewritesToVisibilityAttribute.php

```diff
@@ -0,0 +1,122 @@
+{
+    /**
+     * @var CollectionFactory
+     */
+    private $productCollectionFactory;
+
+    /**
+     * @var ProductUrlRewriteGenerator
+     */
+    private $urlRewriteGenerator;
+
+    /**
+     * @var UrlPersistInterface
+     */
+    private $urlPersist;
+
+    /**
+     * @var ProductUrlPathGenerator
+     */
+    private $urlPathGenerator;
+
+    /**
+     * @param CollectionFactory $collectionFactory
+     * @param ProductUrlRewriteGenerator $urlRewriteGenerator
+     * @param UrlPersistInterface $urlPersist
+     * @param ProductUrlPathGenerator|null $urlPathGenerator
+     */
+    public function __construct(
+        CollectionFactory $collectionFactory,
+        ProductUrlRewriteGenerator $urlRewriteGenerator,
+        UrlPersistInterface $urlPersist,
+        ProductUrlPathGenerator $urlPathGenerator
+    ) {
+        $this->productCollectionFactory = $collectionFactory;
+        $this->urlRewriteGenerator = $urlRewriteGenerator;
+        $this->urlPersist = $urlPersist;
+        $this->urlPathGenerator = $urlPathGenerator;
+    }
+
+    /**
+     * @param array $productIds
+     * @param int $visibility
+     * @throws UrlAlreadyExistsException
+     */
+    public function execute(array $productIds, int $visibility): void
+    {
+        $products = $this->getProductsByIds($productIds);
+
+        /** @var Product $product */
+        foreach ($products as $product) {
+            if ($visibility == Visibility::VISIBILITY_NOT_VISIBLE) {
+                $this->urlPersist->deleteByData(
+                    [
+                        UrlRewrite::ENTITY_ID => $product->getId(),
+                        UrlRewrite::ENTITY_TYPE => ProductUrlRewriteGenerator::ENTITY_TYPE,
+                    ]
+                );
+            } elseif ($visibility !== Visibility::VISIBILITY_NOT_VISIBLE) {
+                $product->setVisibility($visibility);
+                $productUrlPath = $this->urlPathGenerator->getUrlPath($product);
+                $productUrlRewrite = $this->urlRewriteGenerator->generate($product);
+                $product->unsUrlPath();
+                $product->setUrlPath($productUrlPath);
+
+                try {
+                    $this->urlPersist->replace($productUrlRewrite);
+                } catch (UrlAlreadyExistsException $e) {
+                    throw new UrlAlreadyExistsException(
+                        __(
+                            'Can not change the visibility of the product with SKU equals "%1". URL key "%2" for specified store already exists.',
+                            $product->getSku(),
+                            $product->getUrlKey()
+                        ),
+                        $e,
+                        $e->getCode(),
+                        $e->getUrls()
+                    );
+                }
+            }
+        }
+    }
+
+    /**
+     * @param array $productIds
+     * @return array
+     */
+    protected function getProductsByIds(array $productIds): array
```

Comment:
Please change access level to private for this method

Reply:
@sivaschenko Hi. Thank you. Could you please check once again?